Add Hyundai/Kia Connect integration

fuatakgun opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
Adding new integration to connect Hyundai and Kia cars using underlying Bluelink and UVO APIs. 

## Type of change
- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
This PR is intended to set the ground work for new integration.
Integration has been actively developed and improved by time as an custom component and it is time to become part of core.
Custom component: https://github.com/fuatakgun/kia_uvo/

- This PR fixes or closes issue: NA
- This PR is related to issue: NA
- Link to documentation pull request: NA

## Checklist
- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum


To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

fuatakgun committed:
Add Hyundai/Kia Connect integration

homeassistant commented:
Hi @fuatakgun, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#62451).

Once you do that we will be able to review and accept this pull request.

Thanks!

fuatakgun committed:
Merge branch 'dev' into hyundai-kia-connect

fuatakgun commented:
ping @cdnninja

fuatakgun committed:
Merge branch 'home-assistant:dev' into hyundai-kia-connect

fuatakgun commented:
As this is my first PR the pipeline is not running all steps. It stops after the first two test and waits for an approval. Can someone kick it please?

frenck commented on homeassistant/components/hyundai_kia_connect/__init__.py:
```diff
@@ -0,0 +1,38 @@
+from .const import DOMAIN
+from .coordinator import HyundaiKiaConnectDataUpdateCoordinator
+
+PLATFORMS: list[str] = ["sensor"]
```
```suggestion
PLATFORMS: = [Platform.SENSOR]
```

frenck commented on homeassistant/components/hyundai_kia_connect/__init__.py:
```diff
@@ -0,0 +1,38 @@
+
+PLATFORMS: list[str] = ["sensor"]
+
+_LOGGER = logging.getLogger(__name__)
```
Seems to be unused?

frenck commented on homeassistant/components/hyundai_kia_connect/__init__.py:
```diff
@@ -0,0 +1,38 @@
+async def async_setup_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> bool:
+    """Set up Hyundai / Kia Connect from a config entry."""
+    """Set up this integration using UI."""
+    if hass.data.get(DOMAIN) is None:
+        hass.data.setdefault(DOMAIN, {})
+
```
```suggestion
```

frenck commented on homeassistant/components/hyundai_kia_connect/__init__.py:
```diff
@@ -0,0 +1,38 @@
+
+    coordinator = HyundaiKiaConnectDataUpdateCoordinator(hass, config_entry)
+    await coordinator.async_refresh()
+    hass.data[DOMAIN][config_entry.entry_id] = coordinator
```
```suggestion
    hass.data.setdefault(DOMAIN, {})[config_entry.entry_id] = coordinator
```

fuatakgun replied:
I think this is less readable, but do not have strong opinion

frenck commented on homeassistant/components/hyundai_kia_connect/__init__.py:
```diff
@@ -0,0 +1,38 @@
+        hass.data.setdefault(DOMAIN, {})
+
+    coordinator = HyundaiKiaConnectDataUpdateCoordinator(hass, config_entry)
+    await coordinator.async_refresh()
```
```suggestion
    await coordinator.async_config_entry_first_refresh()
```

frenck commented on homeassistant/components/hyundai_kia_connect/__init__.py:
```diff
@@ -0,0 +1,38 @@
+        config_entry, PLATFORMS
+    )
+    if unload_ok:
+        hass.data[DOMAIN] = None
```
This seems incorrect, if one set up 2 config entries/instances of the integration, it would kill both.

fuatakgun replied:
fair callout, will fix in new revision

frenck commented on homeassistant/components/hyundai_kia_connect/config_flow.py:
```diff
@@ -0,0 +1,95 @@
+        except InvalidAuth:
+            errors["base"] = "invalid_auth"
+        except Exception:  # pylint: disable=broad-except
+            _LOGGER.exception("Unexpected exception")
```
This is already logged in the UI?

frenck commented on homeassistant/components/hyundai_kia_connect/translations/en.json:
```diff
@@ -0,0 +1,26 @@
+{
+    "title": "Hyundai / Kia Connect",
```
This file doesn't match `strings.json`? Please update the translations using the translation tools.

frenck commented on homeassistant/components/hyundai_kia_connect/manifest.json:
```diff
@@ -0,0 +1,13 @@
+  "config_flow": true,
+  "documentation": "https://www.home-assistant.io/integrations/hyundai_kia_connect",
+  "requirements": ["hyundai_kia_connect_api==1.7.0"],
+  "ssdp": [],
+  "zeroconf": [],
+  "homekit": {},
```
```suggestion
```

frenck commented on homeassistant/components/hyundai_kia_connect/sensor.py:
```diff
@@ -0,0 +1,28 @@
+    @property
+    def name(self):
+        """Return a name to use for this sensor."""
+        _LOGGER.debug(f"{DOMAIN} - sensor name - {self.coordinator.data}")
```
This seems unneeded to log on accessing a name?

frenck commented on homeassistant/components/hyundai_kia_connect/strings.json:
```diff
@@ -0,0 +1,21 @@
+  "config": {
+    "step": {
+      "user": {
+        "data": {
```
region and brand missing?

frenck commented on homeassistant/components/hyundai_kia_connect/strings.json:
```diff
@@ -0,0 +1,21 @@
+      }
+    },
+    "error": {
+      "cannot_connect": "[%key:common::config_flow::error::cannot_connect%]",
```
I don't see "cannot_connect" in the config flow, is this stale?

frenck commented on homeassistant/components/hyundai_kia_connect/entity.py:
```diff
@@ -0,0 +1,28 @@
+    @property
+    def unique_id(self):
+        """Return a unique ID to use for this entity."""
+        return self.config_entry.entry_id
```
- Don't we have a unique entry of the car?
- What if we have multiple entities of the same platform, how does this make the entity unique per entity?

fuatakgun replied:
you are right about unique ID, fixed. regarding multiple entities, not in scope of initial integration.

fuatakgun replied:
should be fixed now

frenck commented on homeassistant/components/hyundai_kia_connect/entity.py:
```diff
@@ -0,0 +1,28 @@
+class HyundaiKiaConnectEntity(CoordinatorEntity):
+    """Class for base entity for Hyundai / Kia Connect integration."""
+
+    def __init__(self, coordinator, config_entry: ConfigEntry):
```
Missing return type
```suggestion
    def __init__(self, coordinator, config_entry: ConfigEntry) -> None:
```

frenck commented on homeassistant/components/hyundai_kia_connect/entity.py:
```diff
@@ -0,0 +1,28 @@
+        self.config_entry = config_entry
+
+    @property
+    def unique_id(self):
```
Typing missing. Please note, you can also use shorthand attributes as described here: <https://developers.home-assistant.io/docs/core/entity#entity-class-or-instance-attributes>

frenck commented on homeassistant/components/hyundai_kia_connect/entity.py:
```diff
@@ -0,0 +1,28 @@
+    def __init__(self, coordinator, config_entry: ConfigEntry):
+        """Initialize."""
+        super().__init__(coordinator)
+        self.config_entry = config_entry
```
The coordinator already has the config entry property, you might be able to leverage that one instead.

frenck commented on homeassistant/components/hyundai_kia_connect/entity.py:
```diff
@@ -0,0 +1,28 @@
+    def device_info(self):
+        """Return device information to use for this entity."""
+        return {
+            "identifiers": {(DOMAIN, self.unique_id)},
```
This is the unique ID of the entity, not the device. Please don't mix those up :)

frenck commented on homeassistant/components/hyundai_kia_connect/coordinator.py:
```diff
@@ -0,0 +1,82 @@
+
+    def __init__(self, hass: HomeAssistant, config_entry: ConfigEntry) -> None:
+        """Initialize."""
+        region = config_entry.data.get(CONF_REGION)
+        brand = config_entry.data.get(CONF_BRAND)
+        username = config_entry.data.get(CONF_USERNAME)
+        password = config_entry.data.get(CONF_PASSWORD)
+        pin = config_entry.data.get(CONF_PIN)
```
I would just drop these, and use them directly below.

frenck commented on homeassistant/components/hyundai_kia_connect/coordinator.py:
```diff
@@ -0,0 +1,82 @@
+
+        self.platforms: Set[str] = set()
+        token = Token(config_entry.data.get(CONF_TOKEN))
+        api = get_implementation_by_region_brand(region, brand, username, password, pin)
```
```suggestion
        api = get_implementation_by_region_brand(
            region=config_entry.data[CONF_REGION],
            brand=config_entry.data[CONF_BRAND],
            username=config_entry.data[CONF_USERNAME],
            password=config_entry.data[CONF_PASSWORD],
            pin=config_entry.data[CONF_PIN],
        )
```

frenck commented on homeassistant/components/hyundai_kia_connect/coordinator.py:
```diff
@@ -0,0 +1,82 @@
+        pin = config_entry.data.get(CONF_PIN)
+
+        self.platforms: Set[str] = set()
+        token = Token(config_entry.data.get(CONF_TOKEN))
```
We don't have to use `get()` here, as we know it will be there.
```suggestion
        token = Token(config_entry.data[CONF_TOKEN])
```

frenck commented on homeassistant/components/hyundai_kia_connect/coordinator.py:
```diff
@@ -0,0 +1,82 @@
+        password = config_entry.data.get(CONF_PASSWORD)
+        pin = config_entry.data.get(CONF_PIN)
+
+        self.platforms: Set[str] = set()
```
```suggestion
        self.platforms: set[str] = set()
```

fuatakgun replied:
hey @frenck , it is giving me this error: `homeassistant/components/hyundai_kia_connect/coordinator.py:38: error: "set" is not subscriptable, use "typing.Set" instead  [misc]`

frenck replied:
Add `from __future__ import annotations` to the top of your file

frenck commented on homeassistant/components/hyundai_kia_connect/coordinator.py:
```diff
@@ -0,0 +1,82 @@
+        return await self.async_update()
+        try:
+            return await self.async_update()
+        except Exception as exception:
+            raise UpdateFailed() from exception
```
This is already handled by the coordinator

frenck commented on homeassistant/components/hyundai_kia_connect/config_flow.py:
```diff
@@ -0,0 +1,95 @@
+    username = user_input[CONF_USERNAME]
+    password = user_input[CONF_PASSWORD]
+    region = user_input[CONF_REGION]
+    pin = user_input.get(CONF_PIN, DEFAULT_PIN)
```
Move the default into the schema instead.

frenck commented on homeassistant/components/hyundai_kia_connect/config_flow.py:
```diff
@@ -0,0 +1,95 @@
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle the initial step."""
+        await self.async_set_unique_id(DOMAIN)
```
Why limit it by domain?

fuatakgun replied:
as a starting point, I prefer to set one instance of integration and add multi car support as later on. but it is definitely on roadmap

frenck replied:
It is more difficult to change that later, as you would have to migrate all existing entities. I would not recommend on doing this, unless absolutely not possible otherwise (which, is not the case as far as I can see)

fuatakgun replied:
will address in future commit, thanks for callout

frenck replied:
What is blocking it to address it in this PR? I don't see the limitation currently, what causes this limitation to be needed?

fuatakgun replied:
I mean, under this PR, in future commit

fuatakgun replied:
should be fixed now

frenck reviewed (changes requested):
👋 @fuatakgun 

I've done an initial pass to get the review process going.

As for your question about why the CI didn't run, that is GitHub's default. You are a first-time contributor (welcome btw!), so it doesn't run automatically unless a reviewer triggers it. However, that isn't needed yet, as there are review comments at this point.

fuatakgun committed:
incorporate PR comments and increase test coverage

fuatakgun committed:
Merge branch 'hyundai-kia-connect' of https://github.com/fuatakgun/core into hyundai-kia-connect

fuatakgun committed:
increase test coverage and bump version

fuatakgun committed:
address PR comments

fuatakgun committed:
add new sensors, increase test coverage

fuatakgun commented:
config_flow is fully tested (100% coverage)

fuatakgun commented:
example screenshots;

- device list
![image](https://user-images.githubusercontent.com/11085566/147334409-f9381560-4e95-450a-8fe4-4767e314073e.png)


- device detail
![image](https://user-images.githubusercontent.com/11085566/147334359-89a54985-62cf-4acd-8701-6ad448bacc12.png)

fuatakgun committed:
add sensor unit and device class, fix test

cdnninja committed:
fix: Corrected mapping to API

fuatakgun committed:
Merge pull request #2 from fuatakgun/cdnninjaPatch

fix: Corrected API Mapping

cdnninja committed:
chore: Bump API version

cdnninja committed:
chore: Bump requirements Hyundai Connect

cdnninja committed:
Update requirements_test_all.txt

fuatakgun committed:
add new sensors and deprecate token

fuatakgun commented:
> 👋 @fuatakgun
> 
> I've done an initial pass to get the review process going.
> 
> As for your question about why the CI didn't run, that is GitHub's default. You are a first-time contributor (welcome btw!), so it doesn't run automatically unless a reviewer triggers it. However, that isn't needed yet, as there are review comments at this point.

hi @frenck , we are ready for another pass.

fuatakgun committed:
Enable strict typing

fuatakgun committed:
Enable strict typing

cdnninja commented:
Here is what this looks like for a Canadian Kia once we bump the API version: 
![image](https://user-images.githubusercontent.com/6373468/148612558-e2c6c9ea-8714-470b-a0d7-333d2c6f86bc.png)

cdnninja committed:
feat: Added Sensors and bump API (#3)

* feat: Added Hood Sensor

* feat: Added Next and Last Service

* Update requirements_all.txt

* Update manifest.json

* Update requirements_test_all.txt

cdnninja committed:
fix: Corrected Sensor Unique ID (#4)

* fix: Corrected unique ID

Name can change if updated in app.  ID shouldn't.  Renaming in mobile app causes duplicate sensors to be created.

* Update binary_sensor.py

frenck commented on homeassistant/components/hyundai_kia_connect/__init__.py:
```diff
@@ -0,0 +1,32 @@
+
+async def async_unload_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    unload_ok = await hass.config_entries.async_unload_platforms(
+        config_entry, PLATFORMS
+    )
+    if unload_ok:
```
```suggestion
    if unload_ok := await hass.config_entries.async_unload_platforms(
        config_entry, PLATFORMS
    ):
```

frenck commented on homeassistant/components/hyundai_kia_connect/__init__.py:
```diff
@@ -0,0 +1,32 @@
+        config_entry, PLATFORMS
+    )
+    if unload_ok:
+        hass.data[DOMAIN][config_entry.unique_id] = None
```
```suggestion
        del hass.data[DOMAIN][config_entry.unique_id]
```

frenck commented on homeassistant/components/hyundai_kia_connect/binary_sensor.py:
```diff
@@ -0,0 +1,206 @@
+        key="engine_is_running",
+        name="Engine",
+        device_class=BinarySensorDeviceClass.POWER,
+        on_icon="mdi:engine",
+        off_icon="mdi:engine-off",
```
When using a device class, we should not set an icon.

```suggestion
```

fuatakgun replied:
@frenck , having a device class enables specific icons, but it is not limited to it. Name of states and trigger events are evolving around device classes too. Rather than `X becomes on`, it is visible to user as `X is Open`. But using door icons do not add any value to this use case. So, if there is no explicit reason not to mix device classes with icons or architectural decisions, I prefer to keep it as it enhances user experience. Otherwise, I prefer to drop device classes and stick these icons.

frenck commented on homeassistant/components/hyundai_kia_connect/binary_sensor.py:
```diff
@@ -0,0 +1,206 @@
+        key="defrost_is_on",
+        name="Defrost",
+        device_class=BinarySensorDeviceClass.HEAT,
+        on_icon="mdi:car-defrost-front",
+        off_icon="mdi:car-defrost-front",
```
```suggestion
```

frenck commented on homeassistant/components/hyundai_kia_connect/binary_sensor.py:
```diff
@@ -0,0 +1,206 @@
+        key="steering_wheel_heater_is_on",
+        name="Steering Wheel Heater",
+        device_class=BinarySensorDeviceClass.HEAT,
+        on_icon="mdi:steering",
+        off_icon="mdi:steering",
```
```suggestion
```

frenck commented on homeassistant/components/hyundai_kia_connect/binary_sensor.py:
```diff
@@ -0,0 +1,206 @@
+        key="back_window_heater_is_on",
+        name="Back Window Heater",
+        device_class=BinarySensorDeviceClass.HEAT,
+        on_icon="mdi:car-defrost-rear",
+        off_icon="mdi:car-defrost-rear",
```
```suggestion
```

frenck commented on homeassistant/components/hyundai_kia_connect/binary_sensor.py:
```diff
@@ -0,0 +1,206 @@
+        key="side_mirror_heater_is_on",
+        name="Side Mirror Heater",
+        device_class=BinarySensorDeviceClass.HEAT,
+        on_icon="mdi:car-side",
+        off_icon="mdi:car-side",
```
```suggestion
```

frenck commented on homeassistant/components/hyundai_kia_connect/binary_sensor.py:
```diff
@@ -0,0 +1,206 @@
+        key="front_left_seat_heater_is_on",
+        name="Front Left Seat Heater",
+        device_class=BinarySensorDeviceClass.HEAT,
+        on_icon="mdi:car-seat-heater",
+        off_icon="mdi:car-seat-heater",
```
```suggestion
```

frenck commented on homeassistant/components/hyundai_kia_connect/binary_sensor.py:
```diff
@@ -0,0 +1,206 @@
+        key="front_right_seat_heater_is_on",
+        name="Front Right Seat Heater",
+        device_class=BinarySensorDeviceClass.HEAT,
+        on_icon="mdi:car-seat-heater",
+        off_icon="mdi:car-seat-heater",
```
```suggestion
```

frenck commented on homeassistant/components/hyundai_kia_connect/binary_sensor.py:
```diff
@@ -0,0 +1,206 @@
+        key="rear_left_seat_heater_is_on",
+        name="Rear Left Seat Heater",
+        device_class=BinarySensorDeviceClass.HEAT,
+        on_icon="mdi:car-seat-heater",
+        off_icon="mdi:car-seat-heater",
```
```suggestion
```

frenck commented on homeassistant/components/hyundai_kia_connect/binary_sensor.py:
```diff
@@ -0,0 +1,206 @@
+        key="rear_right_seat_heater_is_on",
+        name="Rear Right Seat Heater",
+        device_class=BinarySensorDeviceClass.HEAT,
+        on_icon="mdi:car-seat-heater",
+        off_icon="mdi:car-seat-heater",
```
```suggestion
```

frenck commented on homeassistant/components/hyundai_kia_connect/binary_sensor.py:
```diff
@@ -0,0 +1,206 @@
+        key="front_left_door_is_open",
+        name="Front Left Door",
+        device_class=BinarySensorDeviceClass.DOOR,
+        on_icon="mdi:car-door",
+        off_icon="mdi:car-door",
```
```suggestion
```

frenck commented on homeassistant/components/hyundai_kia_connect/binary_sensor.py:
```diff
@@ -0,0 +1,206 @@
+        key="front_right_door_is_open",
+        name="Front Right Door",
+        device_class=BinarySensorDeviceClass.DOOR,
+        on_icon="mdi:car-door",
+        off_icon="mdi:car-door",
```
```suggestion
```

frenck commented on homeassistant/components/hyundai_kia_connect/binary_sensor.py:
```diff
@@ -0,0 +1,206 @@
+        key="back_left_door_is_open",
+        name="Back Left Door",
+        device_class=BinarySensorDeviceClass.DOOR,
+        on_icon="mdi:car-door",
+        off_icon="mdi:car-door",
```
```suggestion
```

frenck commented on homeassistant/components/hyundai_kia_connect/binary_sensor.py:
```diff
@@ -0,0 +1,206 @@
+        key="back_right_door_is_open",
+        name="Back Right Door",
+        device_class=BinarySensorDeviceClass.DOOR,
+        on_icon="mdi:car-door",
+        off_icon="mdi:car-door",
```
```suggestion
```

frenck commented on homeassistant/components/hyundai_kia_connect/binary_sensor.py:
```diff
@@ -0,0 +1,206 @@
+        key="trunk_is_open",
+        name="Trunk",
+        device_class=BinarySensorDeviceClass.DOOR,
+        on_icon="mdi:car-back",
+        off_icon="mdi:car-back",
```
```suggestion
```

frenck commented on homeassistant/components/hyundai_kia_connect/binary_sensor.py:
```diff
@@ -0,0 +1,206 @@
+        key="hood_is_open",
+        name="Hood",
+        device_class=BinarySensorDeviceClass.DOOR,
+        on_icon="mdi:car",
+        off_icon="mdi:car",
```
```suggestion
```

frenck commented on homeassistant/components/hyundai_kia_connect/binary_sensor.py:
```diff
@@ -0,0 +1,206 @@
+        key="fuel_level_is_low",
+        name="Fuel Low Level",
+        device_class=BinarySensorDeviceClass.LIGHT,
+        on_icon="mdi:gas-station-off",
+        off_icon="mdi:gas-station",
```
```suggestion
```

frenck commented on homeassistant/components/hyundai_kia_connect/binary_sensor.py:
```diff
@@ -0,0 +1,206 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+@dataclass
+class HyundaiKiaBinarySensorEntityDescription(BinarySensorEntityDescription):
+    """A class that describes custom binary sensor entities."""
+
+    on_icon: str | None = None
+    off_icon: str | None = None
+
+
```
```suggestion
```

All cases setting icons use device classes, thus should not set icons at all. Making this obsolete.

frenck commented on homeassistant/components/hyundai_kia_connect/binary_sensor.py:
```diff
@@ -0,0 +1,206 @@
+)
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
```
```suggestion
async def async_setup_entry(
    hass: HomeAssistant,
    entry: ConfigEntry,
    async_add_entities: AddEntitiesCallback,
) -> None:
```

frenck commented on homeassistant/components/hyundai_kia_connect/binary_sensor.py:
```diff
@@ -0,0 +1,206 @@
+                    HyundaiKiaConnectBinarySensor(coordinator, description, vehicle)
+                )
+    async_add_entities(entities, True)
+    return True
```
This method has no return type.

```suggestion
```

frenck commented on homeassistant/components/hyundai_kia_connect/binary_sensor.py:
```diff
@@ -0,0 +1,206 @@
+                entities.append(
+                    HyundaiKiaConnectBinarySensor(coordinator, description, vehicle)
+                )
+    async_add_entities(entities, True)
```
Since we are using a data update coordinator, we already have updated data...

```suggestion
    async_add_entities(entities)
```

frenck commented on homeassistant/components/hyundai_kia_connect/binary_sensor.py:
```diff
@@ -0,0 +1,206 @@
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up binary_sensor platform."""
+    coordinator = hass.data[DOMAIN][config_entry.unique_id]
+    entities = []
```
```suggestion
    entities: list[HyundaiKiaConnectBinarySensor] = []
```

frenck commented on homeassistant/components/hyundai_kia_connect/binary_sensor.py:
```diff
@@ -0,0 +1,206 @@
+    for vehicle_id in coordinator.data.vehicles.keys():
+        vehicle: Vehicle = coordinator.data.vehicles[vehicle_id]
+        for description in SENSOR_DESCRIPTIONS:
+            if getattr(vehicle, description.key, None) is not None:
```
What does `None` mean in this context?
Currently not available? Not available ever?
Could it be `None` because of a temporary communication error?

fuatakgun replied:
None means sensor is not viable for this device.
as an example, EV sensors are not available in combustion engine vehicles or low fuel light is missing from EV cars.
If these data points are missing, they are none.

frenck commented on homeassistant/components/hyundai_kia_connect/binary_sensor.py:
```diff
@@ -0,0 +1,206 @@
+class HyundaiKiaConnectBinarySensor(BinarySensorEntity, HyundaiKiaConnectEntity):
+    """Hyundai / Kia Connect binary sensor class."""
+
+    def __init__(
+        self, coordinator, description: BinarySensorEntityDescription, vehicle: Vehicle
+    ):
```
```suggestion
    def __init__(
        self,
        coordinator: HyundaiKiaConnectDataUpdateCoordinator,
        description: BinarySensorEntityDescription,
        vehicle: Vehicle,
    ) -> None:
```

frenck commented on homeassistant/components/hyundai_kia_connect/binary_sensor.py:
```diff
@@ -0,0 +1,206 @@
+    return True
+
+
+class HyundaiKiaConnectBinarySensor(BinarySensorEntity, HyundaiKiaConnectEntity):
+    """Hyundai / Kia Connect binary sensor class."""
+
```
```suggestion
class HyundaiKiaConnectBinarySensor(BinarySensorEntity, HyundaiKiaConnectEntity):
    """Hyundai / Kia Connect binary sensor class."""

    coordinator: HyundaiKiaConnectDataUpdateCoordinator

```

frenck commented on homeassistant/components/hyundai_kia_connect/binary_sensor.py:
```diff
@@ -0,0 +1,206 @@
+        self, coordinator, description: BinarySensorEntityDescription, vehicle: Vehicle
+    ):
+        """Initialize the sensor."""
+        HyundaiKiaConnectEntity.__init__(self, coordinator, vehicle)
```
```suggestion
        super().__init__(self, coordinator, vehicle)
```

frenck commented on homeassistant/components/hyundai_kia_connect/binary_sensor.py:
```diff
@@ -0,0 +1,206 @@
+    ):
+        """Initialize the sensor."""
+        HyundaiKiaConnectEntity.__init__(self, coordinator, vehicle)
+        self._description = description
```
```suggestion
        self.entity_description = description
```

frenck commented on homeassistant/components/hyundai_kia_connect/binary_sensor.py:
```diff
@@ -0,0 +1,206 @@
+        """Initialize the sensor."""
+        HyundaiKiaConnectEntity.__init__(self, coordinator, vehicle)
+        self._description = description
+        self._key = self._description.key
```
Not needed, as we have the entity description stored already.
```suggestion
```

frenck commented on homeassistant/components/hyundai_kia_connect/binary_sensor.py:
```diff
@@ -0,0 +1,206 @@
+        HyundaiKiaConnectEntity.__init__(self, coordinator, vehicle)
+        self._description = description
+        self._key = self._description.key
+        self._attr_unique_id = f"{DOMAIN}_{vehicle.id}_{self._key}"
```
```suggestion
        self._attr_unique_id = f"{DOMAIN}_{vehicle.id}_{description.key}"
```

frenck commented on homeassistant/components/hyundai_kia_connect/binary_sensor.py:
```diff
@@ -0,0 +1,206 @@
+        self._description = description
+        self._key = self._description.key
+        self._attr_unique_id = f"{DOMAIN}_{vehicle.id}_{self._key}"
+        self._attr_name = f"{vehicle.name} {self._description.name}"
```
```suggestion
        self._attr_name = f"{vehicle.name} {description.name}"
```

frenck commented on homeassistant/components/hyundai_kia_connect/binary_sensor.py:
```diff
@@ -0,0 +1,206 @@
+        self._key = self._description.key
+        self._attr_unique_id = f"{DOMAIN}_{vehicle.id}_{self._key}"
+        self._attr_name = f"{vehicle.name} {self._description.name}"
+        self._attr_device_class = self._description.device_class
```
No need to set this, as we already set the entity description (which is queried in that case).
```suggestion
```

frenck commented on homeassistant/components/hyundai_kia_connect/binary_sensor.py:
```diff
@@ -0,0 +1,206 @@
+        self._attr_unique_id = f"{DOMAIN}_{vehicle.id}_{self._key}"
+        self._attr_name = f"{vehicle.name} {self._description.name}"
+        self._attr_device_class = self._description.device_class
+        if self._key == "data":
+            self._attr_extra_state_attributes = vehicle.data
```
Why are we setting all data as state attributes?

I suggest we don't do this and, in case they are interesting, create entities instead.

```suggestion
```

fuatakgun replied:
this is needed for debugging purposes, as there are mild differences in API responses between different cars, regions and brands. I would keep this on a disabled state and users can enable it if needed. Unless, this is something against `best practices` or `architectural principles`

frenck replied:
> this is needed for debugging purposes

Then we should use debug statements, and not pollute our state engine.

fuatakgun replied:
debug statements are hard to receive for a newbie to home assistant and it includes to touch configuration.yaml to enable debugging. But, this sensor keeps the API response and it is always accessible. Our experience while working on multi-region, multi-car solution proved that having this ease the overall debugging easier. I propose to keep this as disabled by default and user can enable it whenever it is needed.

Moreover, some users notice that their cars supports different sensors and they can easily point out the sensor while sharing their data.

fuatakgun replied:
anyway, to make progress, I am removing this.

frenck commented on homeassistant/components/hyundai_kia_connect/binary_sensor.py:
```diff
@@ -0,0 +1,206 @@
+    @property
+    def is_on(self) -> bool | None:
+        """Return true if the binary sensor is on."""
+        return getattr(self.vehicle, self._key)
```
Instead of using `getattr` add a lambda to the entity description and call that.

fuatakgun replied:
Thanks for through review, i am very much interested on this. What is the underlying reason?

frenck replied:
`getattr` returns an `Any` type. By using lambas we can remove this magic and type it correctly as well.

fuatakgun replied:
any example on this? and can you please mark this `needs to solved` or `nice to have`?

fuatakgun replied:
taking this back :) lambda will enable typed access to variables, better, thanks

frenck commented on homeassistant/components/hyundai_kia_connect/binary_sensor.py:
```diff
@@ -0,0 +1,206 @@
+        """Return true if the binary sensor is on."""
+        return getattr(self.vehicle, self._key)
+
+    @property
+    def icon(self):
+        """Return the icon to use in the frontend, if any."""
+        if (self._description.on_icon == self._description.off_icon) is None:
+            return BinarySensorEntity.icon
+        return self._description.on_icon if self.is_on else self._description.off_icon
```
No longer needed as per review comments above.

```suggestion
```

fuatakgun replied:
Answered in first comment why I want to keep this.

frenck commented on homeassistant/components/hyundai_kia_connect/entity.py:
```diff
@@ -0,0 +1,23 @@
+
+    def __init__(self, coordinator, vehicle):
+        """Initialize the base entity."""
+        CoordinatorEntity.__init__(self, coordinator)
```
```suggestion
        super().__init__(self, coordinator)
```

frenck commented on homeassistant/components/hyundai_kia_connect/sensor.py:
```diff
@@ -0,0 +1,117 @@
+"""Sensor for Hyundai / Kia Connect integration."""
```
Let's remove the sensor platform for now and get the integration merged with a single platform (binary sensor) first.

Additional platforms can be added in follow up PRs

cdnninja committed:
Update homeassistant/components/hyundai_kia_connect/__init__.py

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

cdnninja committed:
Update homeassistant/components/hyundai_kia_connect/__init__.py

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

fuatakgun committed:
address PR comments

fuatakgun committed:
address PR comments

fuatakgun committed:
address PR comments

fuatakgun commented:
@frenck we are ready for 3rd round

fuatakgun committed:
version bump

fuatakgun committed:
version bump

fuatakgun committed:
version bump

fuatakgun commented:
> @frenck we are ready for 3rd round

Ping @frenck

frenck commented:
There are still open comments @fuatakgun

fuatakgun commented:
> There are still open comments @fuatakgun

Open comments require a bit discussion about further steps mainly around overriding icons and using device classes.

cdnninja committed:
fix:  Corrected unique IDs and other feedback items (#5)

* fix:  Corrected unique IDs

* Update binary_sensor.py

* Update binary_sensor.py

* Update entity.py

cdnninja committed:
Update binary_sensor.py

cdnninja commented:
We have corrected a few feedback items we had missed.  Any chance of leaving icons in?

Sander0542 commented on homeassistant/components/hyundai_kia_connect/entity.py:
```diff
@@ -0,0 +1,23 @@
+    @property
+    def device_info(self):
+        """Return device information to use for this entity."""
+        return {
+            "identifiers": {(DOMAIN, self.vehicle.id)},
+            "manufacturer": f"{BRANDS[self.coordinator.vehicle_manager.brand]} {REGIONS[self.coordinator.vehicle_manager.region]}",
+            "model": self.vehicle.model,
+            "name": self.vehicle.name,
+        }
```
This should use the `DeviceInfo` class

https://github.com/home-assistant/core/blob/c0e9cfedfb584d00c525546285b16051c2b089c7/homeassistant/helpers/entity.py#L170

```suggestion
        return DeviceInfo(
            identifiers={(DOMAIN, self.vehicle.id)},
            manufacturer=f"{BRANDS[self.coordinator.vehicle_manager.brand]} {REGIONS[self.coordinator.vehicle_manager.region]}",
            model=self.vehicle.model,
            name=self.vehicle.name,
        )
```

Sander0542 commented on homeassistant/components/hyundai_kia_connect/manifest.json:
```diff
@@ -0,0 +1,10 @@
+  "config_flow": true,
+  "documentation": "https://www.home-assistant.io/integrations/hyundai_kia_connect",
+  "requirements": ["hyundai_kia_connect_api==1.18.12"],
+  "dependencies": [],
```
Empty arrays should be removed

```suggestion
```

Sander0542 commented on homeassistant/components/hyundai_kia_connect/manifest.json:
```diff
@@ -0,0 +1,10 @@
+  "name": "Hyundai / Kia Connect",
+  "config_flow": true,
+  "documentation": "https://www.home-assistant.io/integrations/hyundai_kia_connect",
+  "requirements": ["hyundai_kia_connect_api==1.18.12"],
```
Currently, version 1.31.0 is released. Should this be updated before the release of this integration?

https://pypi.org/project/hyundai-kia-connect-api/1.31.0/

Sander0542 reviewed (changes requested):
I am not able to test this integration yet. However, I have reviewed the code. It looks good, but needs some adjustments before this can be merged.
